### PR TITLE
Add deadlettered messages detector

### PR DESCRIPTION
**Pull request by Davidh-claranet**
I added a detector for the Deadlettered messages metric.
It is my first detector from scratch.
Advices are welcome.
David.

### CODE REVIEW

**Review by xp-1000 — COMMENTED** [review]
Hello, thanks for the contribution.

My first advices would be to rebase your branch from master to get last updates (required for merge) and to use the yaml generator instead of writing the code manually this should ease your work to create a full compliant detector in addition to the review.

check this documentation https://github.com/claranet/terraform-signalfx-detectors/blob/master/docs/development.md#tldr for more information. In short, you just have to:

1. create a new `[0-9][0-9]-mydetector.yaml` file into the `conf` dir of the exiting module which will be used after to automatically generate the changes. following configuration available here : https://github.com/claranet/terraform-signalfx-detectors/blob/master/scripts/templates/values.yaml (sort of helm chart).
1. use options available on https://github.com/claranet/terraform-signalfx-detectors/blob/master/scripts/templates/values.yaml to update this yaml import equivalent configuration than your current code
1. remove your manual changes
1. run docker and then enter in dev env with `make`
1. run `make update-module *azure-service-bus*`
1. commit and push and the CI should be happy

File: modules/integration_azure-service-bus/detectors-service-bus.tf
Comment (xp-1000) [inline modules/integration_azure-service-bus/detectors-service-bus.tf:188]: do not reverse the major condition in the critical one will lead to duplicated alerts when critical threshold is reached.

**Review by xp-1000 — CHANGES_REQUESTED** [review]

File: modules/integration_azure-service-bus/detectors-service-bus.tf
Comment (xp-1000) [inline modules/integration_azure-service-bus/detectors-service-bus.tf:189]: double publish functions here

**Review by xp-1000 — CHANGES_REQUESTED** [review]
your fork still require to be updated

File: modules/integration_azure-service-bus/detectors-service-bus.tf
Comment (Davidh-claranet) [reply modules/integration_azure-service-bus/detectors-service-bus.tf]: Sorry, a copy/paste error....

**Review by xp-1000 — APPROVED** [review]
LGTM, again .. you have to refresh your fork.

waiting for azure owners (@BzSpi or @Shr3ps) reviews before to merge

File: modules/integration_azure-service-bus/detectors-service-bus.tf
Comment (BzSpi) [inline modules/integration_azure-service-bus/detectors-service-bus.tf:179]:
```suggestion
  name = format("%s %s", local.detector_name_prefix, "Azure Service Bus deadlettered messages count")
```

File: modules/integration_azure-service-bus/variables.tf
Comment (BzSpi) [inline modules/integration_azure-service-bus/variables.tf:310]:
```suggestion
  description = "Notification recipients list per severity overridden for deadlettered_messages detector"
```

File: modules/integration_azure-service-bus/variables.tf
Comment (BzSpi) [inline modules/integration_azure-service-bus/variables.tf:304]:
```suggestion
  description = "Disable critical alerting rule for deadlettered_messages detector"
```

File: modules/integration_azure-service-bus/variables.tf
Comment (BzSpi) [inline modules/integration_azure-service-bus/variables.tf:298]:
```suggestion
  description = "Disable critical alerting rule for deadlettered_messages detector"
```

File: modules/integration_azure-service-bus/variables.tf
Comment (BzSpi) [inline modules/integration_azure-service-bus/variables.tf:292]:
```suggestion
  description = "Disable all alerting rules for deadlettered_messages detector"
```

File: modules/integration_azure-service-bus/variables.tf
Comment (BzSpi) [inline modules/integration_azure-service-bus/variables.tf:318]:
```suggestion
  default     = ".mean(by=[' entityname ', 'azure_resource_name', 'azure_resource_group_name', 'azure_region', 'entityname'])"
```

**Review by BzSpi — CHANGES_REQUESTED** [review]
A relative check (number of deadlettered messages compared to the overall messages count) may be a better implementation IMHO.

File: modules/integration_azure-service-bus/detectors-service-bus.tf
Comment (Davidh-claranet) [reply modules/integration_azure-service-bus/detectors-service-bus.tf:179]: Done

File: modules/integration_azure-service-bus/variables.tf
Comment (Davidh-claranet) [reply modules/integration_azure-service-bus/variables.tf:304]: Done

File: modules/integration_azure-service-bus/variables.tf
Comment (Davidh-claranet) [reply modules/integration_azure-service-bus/variables.tf:292]: Done

File: modules/integration_azure-service-bus/variables.tf
Comment (Davidh-claranet) [reply modules/integration_azure-service-bus/variables.tf:318]: Done

**Comment by Davidh-claranet** [comment]
I did the requested changes and the detector signal is now a rate of the deadletteredmessage/messages (percentage)

**Comment by xp-1000** [comment]
ping @BzSpi for a review

**Comment by Davidh-claranet** [comment]
Please do not validate this pull request
I have to do changes.....

**Comment by Davidh-claranet** [comment]
the detector signal is now the deadletteredmessage count

**Comment by xp-1000** [comment]
> Please do not validate this pull request
> I have to do changes.....

ok @Davidh-claranet I let you make the PR ready when it is good for you, I convert it as draft for now

**Comment by Davidh-claranet** [comment]
This PR is ready for merge....

**Comment by Davidh-claranet** [comment]
ok pour merger !

**Review by Shr3ps — APPROVED** [review]
Seems ok, too bad this is not created via yml conf file

**Review by xp-1000 — CHANGES_REQUESTED** [review]
sadly this PR is a bit old and the hcl code to create detectors is not up to date anymore because templating has evolved in previous release published since the PR creation.

this is why it is better to use the generator which will take care of updating code to always match templating rules, especially useful for long running PR like this one.

**Comment by xp-1000** [comment]
> sadly this PR is a bit old and the hcl code to create detectors is not up to date anymore because templating has evolved in previous release published since the PR creation.
> 
> this is why it is better to use the generator which will take care of updating code to always match templating rules, especially useful for long running PR like this one.

see https://github.com/claranet/terraform-signalfx-detectors/pull/345 for implementation with generator and what was missing here for your information (transformation function, lasting ..)

**Pull request closed without merging**
